node/services.js

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var errors = require('./errors');
+
+function TChannelServices() {
+    var self = this;
+    self.services = Object.create(null);
+    self.maxPendingForService = Infinity;
+    self.maxPending = Infinity;
+    self.pending = 0;
+}
+
+TChannelServices.prototype.considerDecline = function considerDecline(req) {
+    var self = this;
+    if (self.pending >= self.maxPending) {
+        return errors.MaxPendingError({
+            pending: self.pending
+        });
+    }
+    if (!req.options.serviceName) {
+        return;
+    }
+    var key = mangle(req.options.serviceName);
+    var service = self.services[key];
+    return service && service.considerDecline();
+};
+
+TChannelServices.prototype.onRequest = function onRequest(req) {
+    var self = this;
+    self.pending++;
+    if (!req.options.serviceName) {
+        return;
+    }
+    var key = mangle(req.options.serviceName);
+    var service = self.services[key];
+    if (!service) {
+        service = new TChannelService();
+        service.serviceName = req.options.serviceName;
+        service.maxPending = self.maxPendingForService;
+        self.services[key] = service;
+    }
+    service.onRequest();
+};
+
+TChannelServices.prototype.onRequestResponse = function onRequestResponse(req) {
+    var self = this;
+    self.pending--;
+    if (!req.options.serviceName) {
+        return;
+    }
+    var key = mangle(req.options.serviceName);
+    var service = self.services[key];
+    service.onRequestResponse();
```

Comment:
yeah, i'll do the defense in depth, just in case we forget to onRequest() first.